File: desktop/src/utils/image_compressor.py
"""
image_compressor.py — student photo utilities.

- Crop to 3:4 aspect, resize max 800x1067, JPEG q80.
- Save under data/foto/<tahun>/<kelas_id>/<nis>.jpg.
- Old photo deleted (no orphans on re-capture).
"""
import os
from PIL import Image

MAX_W, MAX_H = 800, 1067
QUALITY = 80

DATA_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "data")


def compress_and_save(src_path: str, tahun_ajaran: str, kelas_id: int, nis: str, ext: str = ".jpg") -> str:
    """Open [src_path], crop to 3:4, resize, save compressed. Returns abs path."""
    dst_dir = os.path.join(DATA_DIR, "foto", tahun_ajaran, str(kelas_id))
    os.makedirs(dst_dir, exist_ok=True)
    dst = os.path.join(dst_dir, f"{nis}{ext}")

    # remove old
    if os.path.exists(dst):
        os.remove(dst)

    with Image.open(src_path) as im:
        cropped = center_crop_34(im)
        scaled = scale_to_fit(cropped)
        scaled.save(dst, "JPEG", quality=QUALITY)
    return dst


def center_crop_34(im: Image.Image) -> Image.Image:
    """Center-crop to a 3:4 portrait aspect ratio."""
    w, h = im.size
    target_ratio = 3 / 4  # width/height — portrait so height is larger
    # We want final W/H = 3/4 (portrait → H > W)
    if w / h > target_ratio:
        # image is wider than 3:4 — crop sides
        new_w = int(h * target_ratio)
        dw = (w - new_w) // 2
        return im.crop((dw, 0, dw + new_w, h))
    else:
        # image is taller than 3:4 — crop top/bottom
        new_h = int(w / target_ratio)
        dh = (h - new_h) // 2
        return im.crop((0, dh, w, dh + new_h))


def scale_to_fit(im: Image.Image) -> Image.Image:
    """Resize keeping aspect; max 800x1067."""
    w, h = im.size
    if w <= MAX_W and h <= MAX_H:
        return im
    ratio = min(MAX_W / w, MAX_H / h, MAX_W / w)  # fit within box
    new_w = int(w * ratio)
    new_h = int(h * ratio)
    return im.resize((new_w, new_h), Image.LANCZOS)


def load_foto(nis: str, tahun_ajaran: str, kelas_id: int) -> Image.Image | None:
    path = os.path.join(DATA_DIR, "foto", tahun_ajaran, str(kelas_id), f"{nis}.jpg")
    return Image.open(path) if os.path.exists(path) else None

File: android/app/src/main/java/com/absenku/utils/ImageCompressor.kt
package com.absenku.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import java.io.File
import java.io.FileOutputStream

/**
 * ImageCompressor — crops-to-3:4, resizes to max 800x1067, JPEG q80.
 *
 * Storage path: <files-dir>/foto/<tahunAjaran>/<kelasId>/<nis>.jpg
 * Old file deleted on re-capture (no orphans).
 */
object ImageCompressor {

    private const val MAX_W = 800   // portrait: width 800 → height ~1067 (3:4)
    private const val QUALITY = 80

    /**
     * Crop [src] to 3:4 aspect (center crop) and resize+compress.
     * Returns absolute path of the saved file.
     */
    fun saveStudentPhoto(context: Context, src: Bitmap, tahunAjaran: String, kelasId: Long, nis: String): String {
        val cropped = centerCrop34(src)
        val scaled = scaleToFit(cropped)
        val dir = File(context.filesDir, "foto/$tahunAjaran/$kelasId").apply { mkdirs() }
        val file = File(dir, "$nis.jpg")
        if (file.exists()) file.delete()
        FileOutputStream(file).use {
            scaled.compress(Bitmap.CompressFormat.JPEG, QUALITY, it)
        }
        return file.absolutePath
    }

    /** Center-crop a bitmap to a 3:4 aspect ratio portrait. */
    private fun centerCrop34(bmp: Bitmap): Bitmap {
        val w = bmp.width
        val targetH = (w * 4.0 / 3.0).toInt()
        val targetW = w
        val srcH = bmp.height
        val srcW = bmp.width
        val scale = maxOf(targetW.toFloat() / srcW, targetH.toFloat() / srcH)
        val scaledW = (srcW * scale).toInt()
        val scaledH = (srcH * scale).toInt()
        val scaled = Bitmap.createScaledBitmap(bmp, scaledW, scaledH, true)
        val dx = (scaledW - targetW) / 2
        val dy = (scaledH - targetH) / 2
        return Bitmap.createBitmap(scaled, dx.coerceAtLeast(0), dy.coerceAtLeast(0), targetW, targetH)
    }

    /** Resize keeping aspect, max width 800 px. */
    private fun scaleToFit(bmp: Bitmap): Bitmap {
        val w = bmp.width
        val h = bmp.height
        if (w <= MAX_W) return bmp
        val ratio = MAX_W.toFloat() / w
        val newW = MAX_W
        val newH = (h * ratio).toInt()
        val matrix = Matrix().apply { setScale(ratio, ratio) }
        return Bitmap.createBitmap(bmp, 0, 0, w, h, matrix, true)
    }

    /** Load a saved student photo (may return null if not found). */
    fun load(context: Context, tahunAjaran: String, kelasId: Long, nis: String): Bitmap? {
        val path = File(context.filesDir, "foto/$tahunAjaran/$kelasId/$nis.jpg")
        return if (path.exists()) BitmapFactory.decodeFile(path.absolutePath) else null
    }

    /** Uri-based picker result → Bitmap (for gallery/cropped pick). */
    fun uriToBitmap(context: Context, uri: Uri): Bitmap? = try {
        context.contentResolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it) }
    } catch (e: Exception) { e.printStackTrace(); null }
}
